app/src/main/java/com/example/francescovalela/androidtutorialwinter2018/MainActivity.kt

```diff
@@ -3,147 +3,54 @@ package com.example.francescovalela.androidtutorialwinter2018
-            lastOperand = calc.recalculate()
-            refreshResultView(lastOperand ?: 0.0)
-            lastOperator = null
-            refreshFormulaView()
-        }
-    }
-
-    private fun undoListener() {
-        if (digitStack.empty()) {
-            calc.undo()
-            lastOperator = null
-            lastOperand = calc.recalculate()
-            refreshFormulaView()
-            refreshResultView(lastOperand ?: 0.0)
-        } else {
-            digitStack.clear()
-            refreshFormulaView()
-        }
-    }
-
-    private fun pushOperator(op: Char) {
-        if (!digitStack.empty()) {
-            val completeDigitValue = getDigitValueFromStack()
-            calc.pushOperand(completeDigitValue)
-            lastOperand = calc.recalculate()
-            lastOperator = op
-            calc.pushOperator(op)
-            digitStack.clear()
-            refreshResultView(lastOperand ?: 0.0)
-        } else if (lastOperand != null) {
-            calc.pushOperator(op)
-        }
-        lastOperator = op
-        refreshFormulaView()
-    }
-
-    private fun getDigitValueFromStack(): Double{
-        var ret = 0.0
-        digitStack.reversed().forEachIndexed { i, d ->
-            ret += d * Math.pow(10.0, (i).toDouble())
-        }
-        return ret
-    }
-
-    private fun refreshFormulaView() {
-        val completeDigitalValue = getDigitValueFromStack()
-        val curr: String = if (digitStack.empty()) "" else {
-            if (completeDigitalValue - getDigitValueFromStack().toLong() == 0.0)
-                completeDigitalValue.toLong().toString() else completeDigitalValue.toString()
-        }
-
-        when {
-            lastOperator != null && lastOperand != null -> {
-                val template = resources.getString(R.string.formula)
-                val tmp: Double = lastOperand ?: 0.0
-                val previous = if (tmp - tmp.toLong() == 0.0)
-                    tmp.toLong().toString() else tmp.toString()
-                formula.text = String.format(template, previous, lastOperator, curr)
-            }
-            lastOperand != null -> {
-                val tmp: Double = lastOperand ?: 0.0
-                val cleanLast: String = if (tmp - tmp.toLong() == 0.0)
-                    tmp.toLong().toString() else tmp.toString()
-                formula.text = cleanLast
-            }
-            else -> formula.text = curr
-        }
-    }
-
-    private fun refreshResultView(value: Double) {
-        result.text = if (value - value.toLong() == 0.0)
-            value.toLong().toString() else value.toString()
-    }
-
-    private fun genericDigitListener(digit: Double){
-        if(lastOperator == null) {
-            calc.clear()
-            lastOperand = null
-        }
-        digitStack.push(digit)
-        refreshFormulaView()
+        btn_clear.setOnClickListener { userInterfaceAdapter.clear() }
+        btn_equals.setOnClickListener{ userInterfaceAdapter.equals() }
```

Comment:
We may still have to explain the ‘setOnClickListener’ portion. Should we consider having one “regular” way of doing it?

Reply:
not sure what you mean by one "regular" way of doing it.